### ISSUE TITLE
uv add generating unstable environments when used with local version identifiers

### ISSUE THREAD

**Issue by itrase** [astral-sh/uv#7461]
On uv version 0.4.10

To reproduce:
`mkdir project_a`
`cd project_a`
`uv init`
`uv add torch==2.4.1+cu124 --index-url https://download.pytorch.org/whl/cu124`
`uv add numpy`

The resulting environment technically works (torch is available for import and uses the correct version of CUDA), but trying to add numpy (or any other package) will give the following error:
```
  × No solution found when resolving dependencies:
  ╰─▶ Because there is no version of torch==2.4.1+cu124 and your project depends on torch==2.4.1+cu124, we can conclude that
      your project's requirements are unsatisfiable.
  help: If this is intentional, run `uv add --frozen` to skip the lock and sync steps.
```

My expectation with `uv add` is that it is intended to result in a reproducible environment, but it doesn't feel like it is the case here. I know that there is significant weirdness around local version identifiers, so perhaps this is intended behavior - but if so, having uv emit a warning would be good.

We can solve this by adding the following to the pyproject.toml:
```
[tool.uv]
extra-index-url = [
    "https://download.pytorch.org/whl/cu124",
]
```

This allows us to continue to `add` packages as normal. However, if I then initialize another project and try to `add` `project_a`, things break again:

`mkdir project_b`
`cd project_b`
`uv init`
`uv add ../project_a --editable`

We get the same error:
```
  × No solution found when resolving dependencies:
  ╰─▶ Because there is no version of torch==2.4.1+cu124 and project-a==0.1.0 depends on torch==2.4.1+cu124, we can conclude
      that project-a==0.1.0 cannot be used.
      And because only project-a==0.1.0 is available and your project depends on project-a, we can conclude that your
      project's requirements are unsatisfiable.
  help: If this is intentional, run `uv add --frozen` to skip the lock and sync steps.
```
Once again, we can solve this by adding the extra-index-url to `project_b`'s `pyproject.toml`, but my expectation is that the extra index should be inherited automatically.

I believe a similar problem exists for things like torch-scatter, which want to use find-links to be installed:
`uv add torch_scatter  -f https://data.pyg.org/whl/torch-2.4.0+cu124.html`
adding a find-links section to the [tool.uv] allows it to proceed, but results in all downstream projects needing that line as well.

**Comment by charliermarsh** [comment]
I think `uv add` should add the index, and we're going to do that as part of some larger changes to how indexes are handled. (I don't think that `project_b` should implicitly use the PyTorch index, but this will become easier to get right when we make the aforementioned changes.)

**Comment by JackismyShephard** [comment]
@charliermarsh I am experiencing what I believe may be a related issue. I have a distributable cli application that uses pytorch 2.5.1+cuda124 as a dependency and is built using uv. If I try to test the cli application using `uv tool install` (in another project) I get: `error: Because there is no version of torch==2.5.1+cu124 and ultimate-rvc==0.1.5 depends on torch==2.5.1+cu124, we can conclude that ultimate-rvc==0.1.5 cannot be used.` Is a fix on the horizon?

**Comment by itrase** [comment]
I'm installing torch in a normal package (not something intended as a tool), but with the updated version of uv, the syntax in the toml that I am using is:

```
requires-python = ">=3.11,<3.12"
dependencies = [
    "torch==2.5.1+cu124",
]

[tool.uv.sources]
torch = { index = "pytorch" }

[[tool.uv.index]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cu124"
explicit = true

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
```

Does your cli application specify the index like this?

**Comment by JackismyShephard** [comment]
@itrase Yes I specify the index just like that. I don't think it should make a difference whether the application exposes a cli entry point or not. However, it should be noted that the package can be installed (from PyPI) as long as I first explicitly install `pip install torch==2.5.1+cu124 torchaudio==2.5.1+cu124 --index-url https://download.pytorch.org/whl/cu124`.

**Comment by charliermarsh** [comment]
@itrase -- Is this solved by using `uv add --default-index`? That writes the index to `pyproject.toml`.

**Comment by itrase** [comment]
yes, thank you!

**Comment by JackismyShephard** [comment]
@itrase @charliermarsh I am still having problems installing my package which is packaged with uv and depending on torch using pip. It only works if I explicitly install torch first using `pip install torch==2.4.1+cu124 torchaudio==2.4.1+cu124 --index-url https://download.pytorch.org/whl/cu124`. But I guess this might be a pip specific issue due to torch not being hosted on pypi?
